### PR TITLE
Feature/snippet search

### PR DESCRIPTION
**Pull request by vrael560**

### CODE REVIEW

File: client/js/components/pages/snippets-index.jsx [hunk 1]
```diff
@@ -1,30 +1,80 @@
...
 import SnippetActions from '../../actions/snippet-actions';
+import SnippetStore from '../../stores/snippet-store';
+import SnippetSearchStore from '../../stores/snippet-search-store';
 
 class SnippetsIndex extends React.Component {
   constructor(props) {
     super(props);
+    this.state = this.getPropsFromStores();
+    this._searchSnippets = this._searchSnippets.bind(this);
+    this.componentDidMount = this.componentDidMount.bind(this);
```
Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: you dont have to bind this

File: client/js/components/pages/snippets-index.jsx [hunk 2]
```diff
@@ -1,30 +1,80 @@
...
+    this.componentWillReceiveProps = this.componentWillReceiveProps.bind(this);
```
Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: you dont have to bind this

File: client/js/components/pages/snippets-index.jsx [hunk 3]
```diff
@@ -1,30 +1,80 @@
...
+    this.componentWillUnmount = this.componentWillUnmount.bind(this);
```
Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: you dont have to bind this

File: client/js/components/pages/snippets-index.jsx [hunk 4]
```diff
@@ -1,30 +1,80 @@
...
+    this.setState(this.getPropsFromStores(this.state, this.context));
+  }
+
+  onSearch() {
+    this.setState(this.getPropsFromSearchStore(this.state, this.context));
+  }
+
+  _searchSnippets (e) {
+    e.preventDefault();
+    SnippetActions.search(this.refs.search.getValue());
```
Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: ```e.target.value``` wont work?

File: client/js/components/pages/snippets-index.jsx [hunk 5]
```diff
@@ -1,30 +1,80 @@
...
     const languages = [{value: 0, label: ''}, {value: 1, label: 'JavaScript'}];
 
     return (
       <PageWrapper>
         <RaisedButton onClick={ () => this.refs.dialog.open()} label="Add new snippet" primary={true}/>
-        <AltContainer store={SnippetStore}>
-          <SnippetsList />
-        </AltContainer>
+        <h2 style={{fontSize: '24px', margin: '20px 0'}}>All snippets:</h2>
+        <form className="snippet-form" onSubmit={this._searchSnippets} style={{float: 'right'}}>
```
Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: why this isn't a separete component ?

File: server/routes/snippets.js [hunk 6]
```diff
@@ -13,6 +13,25 @@ router.get('/', function (req, res) {
   });
 });
 
+router.get('/search', function (req, res) {
+  var options = {};
+  if (req.query.name) {
+    options.where = { name: req.query.name };
+  }
+  models.Snippet.findAll(options).then(function (snippets) {
```
Comment (blimusiek) [inline server/routes/snippets.js]: @vrael560 you are loading snippets without content and versions. Please use ```withVersions``` scope and serialize objects with ```toJson```

File: client/test/pages/snipetts-index.test.js [hunk 7]
```diff
@@ -1,30 +0,0 @@
-// /*eslint-disable no-unused-vars*/
```
Comment (blimusiek) [inline client/test/pages/snipetts-index.test.js]: and what with that test ?

**Comment by blimusiek** [comment]
@vrael560 your tests are failing
SUMMARY:
✔ 4 tests completed
✖ 2 tests failed

File: client/js/components/pages/snippets-index.jsx (hunk 4 above)
Comment (vrael560) [inline client/js/components/pages/snippets-index.jsx]: it's form submit event

File: client/test/pages/snipetts-index.test.js (hunk 7 above)
Comment (vrael560) [inline client/test/pages/snipetts-index.test.js]: for now I've skipped this cause I've made test for composing components. We have to think if page tests are needed.

**Comment by vrael560** [comment]
@blimusiek

File: client/js/components/pages/snippets-index.jsx [hunk 8]
```diff
@@ -1,35 +1,70 @@
...
+  getPropsFromSearchStore() {
+    return SnippetSearchStore.getState();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState(this.getPropsFromStores(nextProps, this.context));
+  }
+
+  componentWillUnmount() {
+    this.storeListeners.forEach(unlisten => unlisten());
```
Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: @vrael560 that method doesn't have method unlisten

File: client/test/actions/snippet-actions.test.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
+/*eslint-disable no-unused-vars*/
+import React from 'react';
+import alt from '../../js/libs/alt';
+import axios from 'axios';
```
Comment (blimusiek) [inline client/test/actions/snippet-actions.test.js]: ```axios``` and ```alt``` are used

File: server/routes/snippets.js [hunk 10]
```diff
@@ -13,6 +13,25 @@ router.get('/', function (req, res) {
   });
 });
 
+router.get('/search', function (req, res) {
+  var options = {};
+  if (req.query.name) {
+    options.where = { name: req.query.name };
+  }
+  models.Snippet.scope('withVersions').findAll(options).then(function (snippets) {
+    var mappedSnippets = snippets.map(function (s) {
```
Comment (blimusiek) [inline server/routes/snippets.js]: @vrael560 please use ```snippet.toJson()```

File: client/js/components/pages/snippets-index.jsx (hunk 8 above)
Comment (vrael560) [inline client/js/components/pages/snippets-index.jsx]: @blimusiek from alt documentation 
`The listen method returns a function that you can use to unsubscribe to store updates.`

```this.storeListeners.push(SnippetStore.listen(this.onChange));
    this.storeListeners.push(SnippetSearchStore.listen(this.onSearch));```

Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: ok, my bad .... but I got error. DO you know why ?

Comment (vrael560) [inline client/js/components/pages/snippets-index.jsx]: error on this method?

Comment (blimusiek) [inline client/js/components/pages/snippets-index.jsx]: yep

**Comment by blimusiek** [comment]
ok for me

**Pull request merged**